### PR TITLE
Fix random crashes on client disconnect

### LINKED ISSUE #87: Crash when disconnecting clients
The crash appears to be random.
image
image

### PR DESCRIPTION
**Pull request by only-a-ptr**
I changed a few things that seemed dangerous to me, but I can't guarantee it will help.

UPDATE: I found an outrageous multithreading bug in disconnect logic (missing lock). This should fix the crashes for good.
Fixes #87

### CODE REVIEW

**Comment by tritonas00** [comment]
This PR makes the client crash on back to menu or exit game
```
[RoR|Networking] Disconnect() disconnecting...
[RoR|Networking] SendThread stopped
[RoR|Networking] Disconnect() sender thread stopped...
[RoR|Networking] NetFatalError(): disconnected: remote side closed the connection
[RoR|Networking] Disconnect() receiver thread stopped...
SWBaseSocket::waitIO() timeout!
```

**Comment by CuriousMike56** [comment]
Related to #1279

**Comment by CuriousMike56** [comment]
Test server currently has 114 total disconnects, no crashes (yet).
But RoR still crashes on disconnect:
```
11:58:32: [RoR|Networking] Disconnect() disconnecting...
11:58:32: [RoR|Networking] SendThread stopped
11:58:32: [RoR|Networking] Disconnect() sender thread stopped...
11:58:32: [RoR|Networking] NetFatalError(): disconnected: remote side closed the connection
11:58:32: [RoR|Networking] Disconnect() receiver thread stopped...
```
I'm guessing it's an issue with RoR itself as the server seems to correctly disconnect the client. (the crash statistic doesn't update):
```
06-04-2019 11:53:29|t227| INFO| User disconnects on request: CuriousMike
06-04-2019 11:53:29|t227| INFO| ScriptEngine: success: removed a 'frameStep' callback: void chatPlugin_ping::frameStep(float)
06-04-2019 11:53:29|t227|VERBO| Disconnecting client ID 113: disconnected on request
06-04-2019 11:53:29|t227| INFO| crash statistic: 31 of 113 deletes crashed
...
...
06-04-2019 11:55:18|t229| INFO| User disconnects on request: CuriousMike
06-04-2019 11:55:18|t229| INFO| ScriptEngine: success: removed a 'frameStep' callback: void chatPlugin_ping::frameStep(float)
06-04-2019 11:55:18|t229|VERBO| Disconnecting client ID 114: disconnected on request
06-04-2019 11:55:18|t229| INFO| crash statistic: 31 of 114 deletes crashed
```

**Comment by tritonas00** [comment]
probably a change introduced with this PR must be applied on the client also

**Comment by tritonas00** [comment]
@only-a-ptr 

instead of `while (true)` in Receiver and Broadcaster doing `while (m_keep_running)` fixes RoR crash on disconnect

File: source/server/broadcaster.cpp [hunk 1]
```diff
@@ -36,50 +36,45 @@ void *StartBroadcasterThread(void *data) {
...
+        pthread_join(m_thread, nullptr);
+    }
 }
 
 void Broadcaster::Thread() {
-    Logger::Log(LOG_DEBUG, "broadcaster m_thread %u owned by client_id %d", ThreadID::getID(), m_client_id);
-    while (m_is_running) {
+    Logger::Log(LOG_DEBUG, "Started broadcaster thread %u owned by client_id %d", ThreadID::getID(), m_client->GetUserId());
+    bool socket_error = false;
+    while (true) {
```
Comment (AnotherFoxGuy) [inline source/server/broadcaster.cpp]:
```suggestion
    while (m_is_running) {
```

suggested by tritonas00 (https://github.com/RigsOfRods/ror-server/pull/90#issuecomment-498025809)

File: source/server/receiver.cpp [hunk 2]
```diff
@@ -33,71 +33,68 @@ void *LaunchReceiverThread(void *data) {
...
-    Logger::Log(LOG_VERBOSE, "UID %d is switching to FLOW", m_id);
+    Logger::Log(LOG_VERBOSE, "UID %d is switching to FLOW", m_client->GetUserId());
 
     // this prevents the socket from hangingwhen sending data
     // which is the cause of threads getting blocked
-    m_socket->set_timeout(60, 0);
-    while (m_is_running) {
-        if (Messaging::ReceiveMessage(m_socket, &type, &source, &streamid, &len, m_dbuffer,
+    m_client->GetSocket()->set_timeout(60, 0);
+    while (true) {
```
Comment (AnotherFoxGuy) [inline source/server/receiver.cpp]:
```suggestion
    while (m_keep_running) {
```
suggested by tritonas00 (https://github.com/RigsOfRods/ror-server/pull/90#issuecomment-498025809)

File: source/server/broadcaster.cpp (hunk 1 above)
Comment (tritonas00) [reply source/server/broadcaster.cpp]: `m_keep_running` not `m_is_running`

File: source/server/receiver.cpp (hunk 2 above)
Comment (tritonas00) [reply source/server/receiver.cpp]: ok i was wrong, we need it only on Broadcaster

File: source/server/broadcaster.cpp [hunk 3]
```diff
@@ -36,50 +36,45 @@ void *StartBroadcasterThread(void *data) {
...
         queue_entry_t msg;
         // define a new scope and use a scope lock
         {
             MutexLocker scoped_lock(m_queue_mutex);
-            while (m_msg_queue.empty() && m_is_running) {
+            while (m_msg_queue.empty()) {
```
Comment (AnotherFoxGuy) [inline source/server/broadcaster.cpp]:
```suggestion
            while (m_msg_queue.empty() && m_keep_running) {
```
https://github.com/RigsOfRods/ror-server/pull/90#issuecomment-498025809

**Comment by CuriousMike56** [comment]
There still is a possibility of a crash:
image

**Comment by only-a-ptr** [comment]
@CuriousMike56 Are you sure that crash comes from this PR? Because it would make sense in upstream, but it doesn't make sense here.

@tritonas00 @AnotherFoxGuy OK, I'll restore that code bit, it's awful but if it works, then it works.

**Comment by tritonas00** [comment]
>EDIT: I'll make one last attempt to be clever about it, if it fails, I'll just apply your suggestion and be done.

with this change if you are going to connect to an `any` server, disconnecting works before selecting a terrain but still crashes when disconnecting after loaded it

**Comment by CuriousMike56** [comment]
image

**Comment by only-a-ptr** [comment]
Interesting. Last time the crash happened in `std::deque::pop_front()`, which may be legitimate if the std::deque is empty at the time. This time it happens in `std::deque::push_back()` which is never legitimate. We have a memory corruption problem.

**Comment by tritonas00** [comment]
why not just try `if (!m_msg_queue.empty())` in https://github.com/RigsOfRods/ror-server/blob/master/source/server/broadcaster.cpp#L137 ?

**Comment by CuriousMike56** [comment]
@only-a-ptr Would running Valgrind affect user performance? The crash happened ~4 days after the server was started and requires an active stream of players to occur.

**Comment by tritonas00** [comment]
maybe this can help https://github.com/d99kris/heapusage

**Comment by tritonas00** [comment]
It works fine and also on the first client connect it says:

` valgrind --tool=memcheck --track-origins=yes bin/rorserver -lan -port 12000 -fg`

```
18-06-2019 16:06:49|t02| INFO|New client: Archlinux (en_gr), using RoR 0.4.8.0-dev-db39d081-dirt, with token DA39A3EE5E6B4B0D3255BFEF95601890AFD80709
==20993== Thread 4:
==20993== Conditional jump or move depends on uninitialised value(s)
==20993==    at 0x16E15B: Sequencer::queueMessage(int, int, unsigned int, char*, unsigned int) (sequencer.cpp:735)
==20993==    by 0x169358: Receiver::Thread() (receiver.cpp:96)
==20993==    by 0x169428: LaunchReceiverThread(void*) (receiver.cpp:31)
==20993==    by 0x4892A91: start_thread (in /usr/lib/libpthread-2.29.so)
==20993==    by 0x4C9DCD2: clone (in /usr/lib/libc-2.29.so)
==20993==  Uninitialised value was created by a heap allocation
==20993==    at 0x4838DEF: operator new(unsigned long) (vg_replace_malloc.c:334)
==20993==    by 0x16D5CA: Sequencer::createClient(SWInetSocket*, RoRnet::UserInfo) (sequencer.cpp:258)
==20993==    by 0x166249: Listener::threadstart() (listener.cpp:230)
==20993==    by 0x166D38: s_lsthreadstart(void*) (listener.cpp:44)
==20993==    by 0x4892A91: start_thread (in /usr/lib/libpthread-2.29.so)
==20993==    by 0x4C9DCD2: clone (in /usr/lib/libc-2.29.so)
==20993==
```

**Comment by tritonas00** [comment]
> @tritonas00 @AnotherFoxGuy OK, I'll restore that code bit, it's awful but if it works, then it works. EDIT: I'll make one last attempt to be clever about it, if it fails, I'll just apply your suggestion and be done.

RoR still crashes sometimes though (even with my solution), mainly when you spawn an actor then Back to menu :unamused:

**Comment by tritonas00** [comment]
@only-a-ptr 
alt text

**Comment by CuriousMike56** [comment]
Can this be merged? There still is a possibility of a crash, but it's generally much more stable and currently I have to manually apply this PR for the prebuilt versions on the repo. Not to mention the fact it's now 7 months old.

**Comment by only-a-ptr** [comment]
@CuriousMike56 Sorry for the wait, I distracted myself with #98

**Pull request merged**